### PR TITLE
[#IOPID-1195] add IO-WEB profile invalid unlock codes alert

### PR DESCRIPTION
**Pull request by arcogabbo**
### List of changes
* [[#IOPID-1195] added alert](https://github.com/pagopa/io-infra/commit/1eed33e4ed6feaf89dac3d03437215d6db6e4aa0)

### Motivation and context
Add an alert that fires when a lot of invalid codes are submitted to IO-WEB profile unlock endpoint

### Type of changes

- [x] Add new resources
- [ ] Update configuration to existing resources
- [ ] Remove existing resources

### Env to apply

- [ ] DEV
- [ ] UAT
- [x] PROD

### Does this introduce a change to production resources with possible user impact?

- [ ] Yes, users may be impacted applying this change
- [ ] No

### Does this introduce an unwanted change on infrastructure? Check terraform plan execution result

- [ ] Yes
- [ ] No

### Other information

---

### How to apply

After PR is approved
1. run deploy pipeline from Azure DevOps [io-platform-iac-projects](https://dev.azure.com/pagopaspa/io-platform-iac-projects/_build?view=folders)
2. select PR branch
3. wait for approval

### CODE REVIEW

File: src/domains/ioweb-app/06_function_ioweb_profile.tf [hunk 1]
```diff
@@ -316,3 +316,38 @@ resource "azurerm_monitor_autoscale_setting" "function_ioweb_profile" {
...
+
+// ----------------------------------------------------
+// Alerts
+// ----------------------------------------------------
+resource "azurerm_monitor_metric_alert" "alert_too_much_invalid_codes_call_on_unlock" {
+  name                = "[${upper(var.domain)} | ${module.function_ioweb_profile.name}] Unexpected number of invalid codes to unlock endpoint"
+  resource_group_name = azurerm_resource_group.ioweb_profile_rg.name
+  scopes              = [module.function_ioweb_profile.id]
+  description         = "A lot of invalid codes submitted to IO-WEB profile unlock functionality"
+  severity            = 3
```
Comment (gquadrati) [inline src/domains/ioweb-app/06_function_ioweb_profile.tf:328]:
```suggestion
  severity            = 1
```

File: src/domains/ioweb-app/06_function_ioweb_profile.tf [hunk 2]
```diff
@@ -316,3 +316,38 @@ resource "azurerm_monitor_autoscale_setting" "function_ioweb_profile" {
...
 }
+
+// ----------------------------------------------------
+// Alerts
+// ----------------------------------------------------
+resource "azurerm_monitor_metric_alert" "alert_too_much_invalid_codes_call_on_unlock" {
+  name                = "[${upper(var.domain)} | ${module.function_ioweb_profile.name}] Unexpected number of invalid codes to unlock endpoint"
+  resource_group_name = azurerm_resource_group.ioweb_profile_rg.name
+  scopes              = [module.function_ioweb_profile.id]
+  description         = "A lot of invalid codes submitted to IO-WEB profile unlock functionality"
```
Comment (gquadrati) [inline src/domains/ioweb-app/06_function_ioweb_profile.tf:327]:
```suggestion
  description         = "Too many of invalid codes submitted to IO-WEB profile unlock functionality"
```

File: src/domains/ioweb-app/06_function_ioweb_profile.tf [hunk 3]
```diff
@@ -316,3 +316,38 @@ resource "azurerm_monitor_autoscale_setting" "function_ioweb_profile" {
...
+  dynamic_criteria {
+    metric_namespace         = "Microsoft.Web/sites"
+    metric_name              = "Http401"
+    aggregation              = "Total"
+    operator                 = "GreaterThan"
+    alert_sensitivity        = "Medium"
+    evaluation_total_count   = 4
+    evaluation_failure_count = 4
+    skip_metric_validation   = false
+  }
```
Comment (gquadrati) [inline src/domains/ioweb-app/06_function_ioweb_profile.tf:345]: I'd go for a more granular check:

```suggestion
  window_size         = "PT15M"
  frequency           = "PT5M"
  auto_mitigate       = false


  # Metric info
  # https://learn.microsoft.com/en-us/azure/azure-monitor/reference/supported-metrics/microsoft-web-sites-metrics
  dynamic_criteria {
    metric_namespace         = "Microsoft.Web/sites"
    metric_name              = "Http401"
    aggregation              = "Total"
    operator                 = "GreaterThan"
    alert_sensitivity        = "Medium"
    evaluation_total_count   = 4
    evaluation_failure_count = 1
    skip_metric_validation   = false
  }
```

**Review by gquadrati — CHANGES_REQUESTED** [review]

File: src/domains/ioweb-app/06_function_ioweb_profile.tf [hunk 4]
```diff
@@ -316,3 +316,85 @@ resource "azurerm_monitor_autoscale_setting" "function_ioweb_profile" {
...
+  // check once every minute(evaluation_frequency)
+  // on the last minute of data(window_duration)
+  evaluation_frequency = "PT1M"
+  window_duration      = "PT1M"
+
+  criteria {
+    query                   = <<-QUERY
+requests
+| where cloud_RoleName == "${module.function_ioweb_profile.name}"
+| where url contains "unlock-session"
```
Comment (BurnedMarshal) [inline src/domains/ioweb-app/06_function_ioweb_profile.tf:342]: We can optimize the query using `==` instead `contains`. [ref](https://learn.microsoft.com/it-it/azure/data-explorer/kusto/query/datatypes-string-operators#performance-tips)
```suggestion
| where name == "UnlockSession"
```

File: src/domains/ioweb-app/06_function_ioweb_profile.tf [hunk 5]
```diff
@@ -316,3 +316,85 @@ resource "azurerm_monitor_autoscale_setting" "function_ioweb_profile" {
...
 }
+
+// ----------------------------------------------------
+// Alerts
+// ----------------------------------------------------
+resource "azurerm_monitor_scheduled_query_rules_alert_v2" "alert_too_much_invalid_codes_on_unlock" {
+  enabled                 = true
+  name                    = "[${upper(var.domain)} | ${module.function_ioweb_profile.name}] Unexpected number of invalid codes to unlock endpoint"
+  resource_group_name     = azurerm_resource_group.ioweb_profile_rg.name
+  scopes                  = [module.function_ioweb_profile.id]
```
[the remark this answers is not in the document]
Comment (arcogabbo) [reply src/domains/ioweb-app/06_function_ioweb_profile.tf:327]: nice suggestion, thank you. done in https://github.com/pagopa/io-infra/pull/828/commits/79c045310d6c3222206cf0b33abee89715c76962

**Review by BurnedMarshal — APPROVED** [review]

**Review by gquadrati — APPROVED** [review]

**Pull request merged**